### ISSUE TITLE
RightCurly: Update logic for option ALONE and ALONE_OR_SINGLELINE to support double brace initializes handling without shouldStartLine

### ISSUE THREAD

**Issue by MEZk** [checkstyle/checkstyle#4057]
RightCurlyCheck should validate double brace initializes without the need of shouldStartLine=true.
For example:
````
$ cat TestClass.java
public class TestClass {
    void method() {
        java.util.Map<String, String> map2 = new java.util.LinkedHashMap<String, String>() {{
            put("Hello", "World");
            put("first", "second");
            put("polygene", "lubricants");
            put("alpha", "betical");}  //violation (line 7)
        };
    }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
         <module name="RightCurly">
             <property name="option" value="alone"/>
             <property name="shouldStartLine" value="false"/>
             <property name="tokens" value="INSTANCE_INIT"/>
         </module>
    </module>
</module>

$ java -jar checkstyle-7.6-all.jar -c TestConfig.xml TestClass.java
Starting audit...
Audit done.
````

If shouldStartLine = true, there will be a violation on line 7 for ALONE and ALONE_OR_SINGLELINE.

http://rveach.no-ip.org/checkstyle/checkstyle.php?action=view&config=d74389219b72d7663eea61765b13cea7&code=2d05a2db84e848faab58bc78208e0d15&checkstyle=checkstyle-7.6-all.jar&printTree=

UPDATE:
Expected: NO violations in both cases. Validation for such cases will be done in new Check - #5945

**Comment by wate123** [comment]
i'm on it

**Comment by ps-sp** [comment]
Rcurly of [this](https://github.com/checkstyle/checkstyle/blob/master/src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/InputRightCurlyAloneOrSingleline.java#L108) double brace initialisation is considered ok ? But `ALONE_OR_SINGLELINE` should enforce `ALONE` or `SINGLELINE` and that `rightCurly` is neither _alone_ nor on the _sameline_ as the `leftCurly`. Why such exception ?

**Comment by ps-sp** [comment]
same goes for [this](https://github.com/checkstyle/checkstyle/blob/master/src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/InputRightCurlyAloneOrSingleline.java#L110). If such behavior is desirable than shouldn't xdocs be updated for this ?

**Comment by MEZk** [comment]
> Why such exception ?

The exception to the general rule is to relax the validation policy for double brace map initializes. The following code example demonstrates that if we put ```}``` alone on the line, the code will become ugly.

```java
        java.util.Map<String, String> map = new java.util.LinkedHashMap<String, String>() {{
            put("Hello", "World");
            put("first", "second");
            put("polygene", "lubricants");
            put("alpha", "betical");
        }
};
```

The main idea of the issues is to reduce the influence of shouldStartLine property on the check's behavior so as to allow to remove the property easily.

**Comment by romani** [comment]
@MEZk and @ps-sp  I updated expectations - there should be no violations.
As we remove this property, this issue can be closed.

**Comment by ps-sp** [comment]
@romani So I guess we need to remove `INSTANCE_INIT` token from this `RightCurlyCheck` and move it to the new check that will be created in #5945. Is that correct ?

**Comment by romani** [comment]
@rnveach , looks like @ps-sp  is right, please look as code below, it is expression like block. When ever compiler force you to write "};" it is expression.

```
$ javac Test.java 
Note: Test.java uses unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.

$ cat Test.java 
import java.util.HashMap;
public class Test {
    void method(Object obj) {
        method(new HashMap() {{
            put("", "");
        }});
    }
}

$ java -jar checkstyle-8.10.1-all.jar -t Test.java
IMPORT -> import [1:0]
|--DOT -> . [1:16]
|   |--DOT -> . [1:11]
|   |   |--IDENT -> java [1:7]
|   |   `--IDENT -> util [1:12]
|   `--IDENT -> HashMap [1:17]
`--SEMI -> ; [1:24]
CLASS_DEF -> CLASS_DEF [2:0]
|--MODIFIERS -> MODIFIERS [2:0]
|   `--LITERAL_PUBLIC -> public [2:0]
|--LITERAL_CLASS -> class [2:7]
|--IDENT -> Test [2:13]
`--OBJBLOCK -> OBJBLOCK [2:18]
    |--LCURLY -> { [2:18]
    |--METHOD_DEF -> METHOD_DEF [3:4]
    |   |--MODIFIERS -> MODIFIERS [3:4]
    |   |--TYPE -> TYPE [3:4]
    |   |   `--LITERAL_VOID -> void [3:4]
    |   |--IDENT -> method [3:9]
    |   |--LPAREN -> ( [3:15]
    |   |--PARAMETERS -> PARAMETERS [3:16]
    |   |   `--PARAMETER_DEF -> PARAMETER_DEF [3:16]
    |   |       |--MODIFIERS -> MODIFIERS [3:16]
    |   |       |--TYPE -> TYPE [3:16]
    |   |       |   `--IDENT -> Object [3:16]
    |   |       `--IDENT -> obj [3:23]
    |   |--RPAREN -> ) [3:26]
    |   `--SLIST -> { [3:28]
    |       |--EXPR -> EXPR [4:14]
    |       |   `--METHOD_CALL -> ( [4:14]
    |       |       |--IDENT -> method [4:8]
    |       |       |--ELIST -> ELIST [4:15]
    |       |       |   `--EXPR -> EXPR [4:15]
    |       |       |       `--LITERAL_NEW -> new [4:15]
    |       |       |           |--IDENT -> HashMap [4:19]
    |       |       |           |--LPAREN -> ( [4:26]
    |       |       |           |--ELIST -> ELIST [4:27]
    |       |       |           |--RPAREN -> ) [4:27]
    |       |       |           `--OBJBLOCK -> OBJBLOCK [4:29]
    |       |       |               |--LCURLY -> { [4:29]
    |       |       |               |--INSTANCE_INIT -> INSTANCE_INIT [4:30]
    |       |       |               |   `--SLIST -> { [4:30]
    |       |       |               |       |--EXPR -> EXPR [5:15]
    |       |       |               |       |   `--METHOD_CALL -> ( [5:15]
    |       |       |               |       |       |--IDENT -> put [5:12]
    |       |       |               |       |       |--ELIST -> ELIST [5:16]
    |       |       |               |       |       |   |--EXPR -> EXPR [5:16]
    |       |       |               |       |       |   |   `--STRING_LITERAL -> "" [5:16]
    |       |       |               |       |       |   |--COMMA -> , [5:18]
    |       |       |               |       |       |   `--EXPR -> EXPR [5:20]
    |       |       |               |       |       |       `--STRING_LITERAL -> "" [5:20]
    |       |       |               |       |       `--RPAREN -> ) [5:22]
    |       |       |               |       |--SEMI -> ; [5:23]
    |       |       |               |       `--RCURLY -> } [6:8]
    |       |       |               `--RCURLY -> } [6:9]
    |       |       `--RPAREN -> ) [6:10]
    |       |--SEMI -> ; [6:11]
    |       `--RCURLY -> } [7:4]
    `--RCURLY -> } [8:0]
```

**Comment by rnveach** [comment]
If this is the case, then you can say method definitions can be inside an expression because they too can also be placed inside an expression block inside an anonymous class.

````
class Test {
    public static void main(String[] args) {
        Runnable runnable = new Runnable() {
            public void run() {
                //
            }
        };
   }
}
````

> When ever compiler force you to write "};" it is expression.

The INSTANCE_INIT is not written like `};`. It only has `}`. The `};` is the anonymous class.

If you truly want to separate how INSTANCE_INIT is handled between inside class and inside anonymous class, then you will have to add the token to both modules.

**Comment by romani** [comment]
Yes , you are right inner brace is part of anonymous class(block brace), outter brace is part of anonymous class that is expression block.

http://checkstyle.sourceforge.net/config_blocks.html#RightCurly shouldStartLine is true by default, so this issue does not make sense ones we remove this property.

@ps-sp , please confirm.

**Comment by rnveach** [comment]
@romani Should INSTANCE_INIT be in both checks? if so, we will need a new issue to ignore expression versions for old check and only print violations when it is part of a class. #5945 would also need this as a token.

**Comment by ps-sp** [comment]
@romani  @rnveach 
> The INSTANCE_INIT is not written like };. It only has }. The }; is the anonymous class.

Yeah that is correct. I was wrong in suggesting we should remove `INSTANCE_INIT` from the existing `RightCurlyCheck`. I didn't think it through properly. So sorry for inconvenience.

>http://checkstyle.sourceforge.net/config_blocks.html#RightCurly shouldStartLine is true by default, so this issue does not make sense ones we remove this property.

Yes. This issue doesn't make any sense in the wake of #5945 . This issue was previously separated out from #4022 and was a blocker for #4022 if I remember correctly, but such validation can now be done in the new check like you said.

**Comment by romani** [comment]
INSTANCE_INIT should stay in this RightCurly as it is not an expression block.
but looks like there we need to allow for `alone` mode to have no violation for
```
        method(new HashMap() {{
            put("", "");
        }}); // allow not-alone ONLY if nothing more is defined.
        HashMap map = new HashMap() {
            {
                put("", "");
            }

            public void clear() {
                // do nothing
            }
        };
```
and do violation on:
```
        HashMap map = new HashMap() {
            {
                put("", "");}

            public void clear() {
                // do nothing
            }
        };
```

@rnveach , why do we need it in new Check ?

**Comment by romani** [comment]
@rnveach ,

in new Check(#5945) we will validate only outer curly brace. Inner curly (INSTANCE_INIT) should stay in this Check.

> What if users don't want it written like this? 

such a rare user, can make its own Check with mode(`absolutely_alone`) and use it. Such allowance to customize Check will cost us a lot and a lot of unclear behavior on conner cases. It is better to keep Check simple and manageable and work for most users.

**Comment by romani** [comment]
not a google style issue any more after introduction of https://github.com/checkstyle/checkstyle/issues/7541
